[TypeReconstruction] Fix reconstructions for allocator.

dcci opened the pull request:
Adrian reported this bug he discovered as part of his
work on improving generics in swift/lldb.
Allocators were treated as `function`, when the demangle
tree suggests they're much more akin to constructors. On
top of that, the code for handling them was commented out,
so type reconstruction was always failing. I'm not entirely
sure this is always correct, but it seems to be an improvement
on what currently type reconstruction does. At least now
we have tests so if something will break, we're going to
notice.

<rdar://problem/38720742>

Commit:
[TypeReconstruction] Fix reconstructions for allocator.

Adrian reported this bug he discovered as part of his
work on improving generics in swift/lldb.
Allocators were treated as `function`, when the demangle
tree suggests they're much more akin to constructors. On
top of that, the code for handling them was commented out,
so type reconstruction was always failing. I'm not entirely
sure this is always correct, but it seems to be an improvement
on what currently type reconstruction does. At least now
we have tests so if something will break, we're going to
notice.

<rdar://problem/38720742>

dcci commented:
@swift-ci please test

jrose-apple commented on lib/IDE/TypeReconstruction.cpp:
```diff
@@ -1347,25 +1347,6 @@ static void VisitNodeFunction(
-  //                            {
-  //                            }
-  //                        }
-  //                    }
```
Did you mean to leave this commented out?

dcci replied:
Yes, because this was wrong code (in the wrong place). I decided to remove it altogether but happy to split it in a separate commit if you prefer.

jrose-apple replied:
Oh, sorry, my brain thought this was an *addition* of commented-out code. I agree that it should be removed.

jrose-apple reviewed (commented):
Yes, allocators are like constructors. :-(

dcci commented:
hmm, this seems to fail on `iphonesimulator-i386`
I'm not sure what's different there, any clue? (or is there an easy way for me to reproduce?)

jrose-apple commented:
That might just mean it doesn't work on *any* simulator.

You can run simulator tests locally if you built with Ninja and not Xcode by using the test-iphonesimulator-* directories instead of test-macosx-x86_64.

dcci commented:
Thanks to Jordan's help I found out this breaks a test on i386-simulator because it always set the default target triple instead of the one of the module. I'm going to fix that.

Commit:
[lldb-moduleimport-test] Pass a target triple as a stopgap solution.

We're going to infer all these informations from the module, but
that's a large refactoring task which I'm going to do next.
In the meanwhile, this should allow Adrian's work to be unblocked.

<rdar://problem/38720742>

dcci commented:
Pushing a stopgap solution to unblock Adrian, and I'll take care of refactoring next

dcci commented:
@swift-ci please test and merge

dcci commented:
@swift-ci please smoke test and merge

dcci commented:
@swift-ci please smoke test macosx

dcci commented:
@shahmishal the linux bot got stuck somehow can you give it a kick?

dcci commented:
@swift-ci plase smoke test linux
